### PR TITLE
Add runbook-run TC server-side

### PR DESCRIPTION
**Pull request by long-sean**
- Add frontend edit and view
- Mandatory param validation
- Test coverage
- Agent skeleton awaiting SDK

### CODE REVIEW

File: octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    return String.format(
+        "Runbook name: %s\nProject name: %s",
+        runbookRunUserData.getRunbookName(), runbookRunUserData.getProjectName());
+  }
+
+  @Override
+  public List<InvalidProperty> validateProperties(Map<String, String> properties) {
+    final List<InvalidProperty> failedProperties = Lists.newArrayList();
+
+    if (StringUtil.isEmpty(properties.getOrDefault(KEYS.getRunbookNamePropertyName(), ""))) {
```
Comment (rain-on) [inline octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java:36]: do we need to do the getOrDefault? I.e. doesn't isEmpty check against null?

File: octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    final String environmentIdentifiers =
+        properties.getOrDefault(KEYS.getEnvironmentNamesPropertyName(), "");
+    if (StringUtil.isEmpty(environmentIdentifiers)) {
+      failedProperties.add(
+          new InvalidProperty(
+              KEYS.getEnvironmentNamesPropertyName(),
+              "At least one environment name must be specified."));
+    } else {
+      List<String> environmentNames = StringUtil.split(environmentIdentifiers, "\n");
+      for (String identifier : environmentNames) {
```
Comment (rain-on) [inline octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java:59]: you can probably do something nice with streams & filter here to make remove the if/else and loop.

File: octopus-server/src/test/java/octopus/teamcity/server/generic/RunbookRunStepTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    assertThat(invalidProperties)
+        .isNotNull()
+        .hasSize(1)
+        .flatExtracting(InvalidProperty::getPropertyName, InvalidProperty::getInvalidReason)
+        .containsExactly(
+            RunbookRunPropertyNames.RUNBOOK_NAME,
+            "Runbook name must be specified and cannot be whitespace.");
+  }
+
+  @Test
```
Comment (rain-on) [inline octopus-server/src/test/java/octopus/teamcity/server/generic/RunbookRunStepTest.java:59]: have you checkouted out @ParameterisedTest - you can probably get many of these tests "for free"

**Review by rain-on — COMMENTED** [review]
LGTM - see if "getOfDefault" is needed - and see if @ParameterisedTest reduces SLOC

File: octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import octopus.teamcity.common.runbookrun.RunbookRunUserData;
+
+public class RunbookRunStep extends OctopusBuildStep {
+
+  private final RunbookRunPropertyNames KEYS = new RunbookRunPropertyNames();
+
+  public RunbookRunStep() {
+    super(
+        "runbook-run",
+        "Runbook run",
```
[the remark this answers is not in the document]
Comment (long-sean) [reply octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java:19]: Have gone with "Run runbook", personally prefer "Execute" but am trying to be consistent with the Octopus terminology

File: octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java (hunk 2 above)
Comment (long-sean) [reply octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java:58]: I cannot see a way to do this that would be simpler/more efficient. If you have ideas I would like to talk through it

File: octopus-server/src/test/java/octopus/teamcity/server/generic/RunbookRunStepTest.java (hunk 3 above)
Comment (long-sean) [reply octopus-server/src/test/java/octopus/teamcity/server/generic/RunbookRunStepTest.java]: vnext... now with parameterised testing

File: octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java (hunk 1 above)
Comment (long-sean) [reply octopus-server/src/main/java/octopus/teamcity/server/generic/RunbookRunStep.java]: Fixed

File: octopus-common/src/main/java/octopus/teamcity/common/runbookrun/RunbookRunUserData.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    return fetchRaw(KEYS.getRunbookNamePropertyName());
+  }
+
+  public String getProjectName() {
+    return fetchRaw(KEYS.getProjectNamePropertyName());
+  }
+
+  public List<String> getEnvironmentNames() {
+    final String rawInput = fetchRaw(KEYS.getEnvironmentNamesPropertyName());
+    return StringUtil.split(rawInput, "\n");
```
[the remark this answers is not in the document]
Comment (long-sean) [reply octopus-common/src/main/java/octopus/teamcity/common/runbookrun/RunbookRunUserData.java]: new method introduced to Base for fetching raw from newline delimited

**Pull request merged**